src/core/cluster.py

```diff
@@ -5,27 +5,45 @@
 """Objects representing the state of KafkaCharm."""
 
 import os
-
+from functools import cached_property
+
+from charms.data_platform_libs.v0.data_interfaces import (
+    DatabaseRequirerData,
+    DataPeerData,
+    DataPeerOtherUnitData,
+    DataPeerUnitData,
+    KafkaProvidesData,
+)
 from ops import Framework, Object, Relation
+from ops.model import Unit
 
-from core.models import KafkaBroker, KafkaCluster, ZooKeeper
+from core.models import SUBSTRATES, KafkaBroker, KafkaClient, KafkaCluster, ZooKeeper
 from literals import (
     INTERNAL_USERS,
     PEER,
     REL_NAME,
+    SECRETS_UNIT,
     SECURITY_PROTOCOL_PORTS,
     ZK,
     Status,
-    Substrate,
 )
 
 
 class ClusterState(Object):
-    """Properties and relations of the charm."""
+    """Collection of global cluster state for the Kafka services."""
 
-    def __init__(self, charm: Framework | Object, substrate: Substrate):
+    def __init__(self, charm: Framework | Object, substrate: SUBSTRATES):
         super().__init__(parent=charm, key="charm_state")
-        self.substrate: Substrate = substrate
+        self.substrate: SUBSTRATES = substrate
+
+        self.peer_app_interface = DataPeerData(self.model, relation_name=PEER)
+        self.peer_unit_interface = DataPeerUnitData(
+            self.model, relation_name=PEER, additional_secret_fields=SECRETS_UNIT
+        )
+        self.zookeeper_requires_interface = DatabaseRequirerData(
+            self.model, relation_name=ZK, database_name=f"/{self.model.app.name}"
```

Comment:
Because `app-name` would be Kafka in both cases you mean?
I would say that it's possible, but we wouldn't recommend or support it in a prod setting, because it'll use up resources.

@deusebio - thoughts here? It's not a problem in non-cross-model relations because in-model app names have to be unique, but is this something we should care about? Or are we happy with not supporting it?

Reply:
I would not say it is very likely. One set up could be having a single ZooKeeper that would be the central control plane for multiple Kafka deployments, but in this case every Kafka deployment would likely have a different scope and/or name. 

The use-case where people may want to keep the same name is if the kafka deployment have the same scope but in prod and pre-prod, but we should not really advocate for this. 

Overall, i would prefer simplicity (otherwise we would be creating zNodes with the model name as well), given the constraints we are proving are not that blocking (they just need to call the charms a bit more specifically, but that would help them understand the zNodes a bit better)